spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/security/IcebergHiveConnectorDelegationTokenProvider.java

```diff
@@ -0,0 +1,240 @@
+    }
+  }
+
+  IMetaStoreClient createHmsClient(HiveConf conf) throws Exception {
+    return new HiveMetaStoreClient(conf, null, false);
+  }
+
+  @Override
+  public Option<Object> obtainDelegationTokens(
+      Configuration hadoopConf, SparkConf sparkConf, Credentials creds) {
+    Map<String, IMetaStoreClient> hmsClientCache = Maps.newHashMap();
+    Set<String> requireTokenCatalogs = getRequireTokenCatalogs(sparkConf);
+    LOG.debug("Require token Hive catalogs: {}", requireTokenCatalogs);
+    try {
+      for (String catalogName : requireTokenCatalogs) {
+        Optional<HiveConf> hiveConfOpt = buildHiveConf(sparkConf, hadoopConf, catalogName);
+        if (!hiveConfOpt.isPresent()) {
+          continue;
+        }
+
+        HiveConf remoteHmsConf = hiveConfOpt.get();
+        String metastoreUri = sparkConf.get(SPARK_SQL_CATALOG_PREFIX + catalogName + URI_KEY);
+        String principal = sparkConf.get(SPARK_SQL_CATALOG_PREFIX + catalogName + PRINCIPAL_KEY);
+        try {
+          UserGroupInformation currentUser = UserGroupInformation.getCurrentUser();
+          LOG.debug(
+              "Getting Hive delegation token for {} against {} at {}",
+              currentUser.getUserName(),
+              principal,
+              metastoreUri);
+          doAsRealUser(
+              () -> {
+                IMetaStoreClient hmsClient = hmsClientCache.get(metastoreUri);
+                if (hmsClient == null) {
+                  hmsClient = createHmsClient(remoteHmsConf);
+                  hmsClientCache.put(metastoreUri, hmsClient);
+                }
+
+                String tokenStr =
+                    hmsClient.getDelegationToken(currentUser.getUserName(), principal);
+                Token<DelegationTokenIdentifier> hive2Token = new Token<>();
+                hive2Token.decodeFromUrlString(tokenStr);
+                LOG.info("Get Token from hive metastore: {}", hive2Token);
+                creds.addToken(new Text(metastoreUri), hive2Token);
+                return null;
+              });
+        } catch (IOException | UndeclaredThrowableException e) {
+          LOG.error(
+              "Failed to obtain delegation token for catalog {}: {}",
+              catalogName,
+              e.getMessage(),
+              e);
+          throw new RuntimeException(e);
+        } catch (Throwable t) {
+          LOG.error(
+              "Unexpected error when obtaining delegation token for catalog {}: {}",
+              catalogName,
+              t.getMessage(),
+              t);
+          throw new RuntimeException(t);
+        }
+      }
+
+      return Option.empty();
+    } catch (NoClassDefFoundError e) {
+      throw new RuntimeException(e);
+    } finally {
+      close(hmsClientCache);
+    }
+  }
+
+  private void close(Map<String, IMetaStoreClient> hmsClientCache) {
+    hmsClientCache
+        .values()
+        .forEach(
+            client -> {
+              try {
+                client.close();
```

Comment:
I'm not 100% sure that clients can be closed all the time, like when it's corrupted. Such case stopping the job by throwing `RuntimeException` is not advisable. I would try to close it silently with some warning.